stellargraph/data/explorer.py

```diff
@@ -107,6 +81,114 @@ def run(self, **kwargs):
+        if nodes is None:
+            self._raise_error("A list of root node IDs was not provided.")
+        if not is_real_iterable(nodes):
+            self._raise_error("Nodes parameter should be an iterable of node IDs.")
+        if (
+            len(nodes) == 0
+        ):  # this is not an error but maybe a warning should be printed to inform the caller
+            print(
+                "({}) WARNING: No root node IDs given. An empty list will be returned as a result.".format(
+                    type(self).__name__
+                )
+            )
+
+    def _check_repetitions(self, n):
+        if type(n) != int:
+            self._raise_error(
+                "The number of walks per root node, n, should be integer type."
+            )
+        if n <= 0:
+            self._raise_error(
+                "The number of walks per root node, n, should be a positive integer."
+            )
+
+    def _check_length(self, length):
+        if type(length) != int:
+            self._raise_error("The walk length, length, should be integer type.")
+        if length <= 0:
+            # Technically, length 0 should be okay, but by consensus is invalid.
+            self._raise_error("The walk length, length, should be a positive integer.")
+
+    def _check_seed(self, seed):
+        if seed is None:
+            # Restore the random state
+            return self._random_state
+        if type(seed) != int:
+            self._raise_error(
+                "The random number generator seed value, seed, should be integer type or None."
+            )
+        if seed < 0:
+            self._raise_error(
+                "The random number generator seed value, seed, should be non-negative integer or None."
+            )
+        # seed the random number generator
+        return random.Random(seed)
+
+    # For neighbourhood sampling
+    def _check_sizes(self, n_size):
+        err_msg = "The neighbourhood size must be a list of non-negative integers."
+        if n_size is None or type(n_size) != list:
+            self._raise_error(err_msg)
+        if len(n_size) == 0:
+            # Technically, length 0 should be okay, but by consensus it is invalid.
+            self._raise_error("The neighbourhood size should not be empty.")
+        for d in n_size:
+            if type(d) != int or d < 0:
+                self._raise_error(err_msg)
+
+
+class HeterogeneousGraphWalk(GraphWalk):
+    """
+    Base class for exploring graphs with heterogeneous edge types.
+    """
+
+    def __init__(self, graph, graph_schema=None, seed=None):
+        super().__init__(graph, graph_schema, seed)
+
+        # Create a dict of adjacency lists per edge type, for faster neighbour sampling from graph in SampledHeteroBFS:
+        # TODO: this could be better placed inside StellarGraph class
+        edge_types = self.graph_schema.edge_types
+        self.adj = {et: defaultdict(lambda: [None]) for et in edge_types}
```

Comment:
If this represents no neighbours, should this be

```suggestion
        self.adj = {et: defaultdict(lambda: []) for et in edge_types}
```

If not, why `[None]` exactly?

Reply:
You will need to ask the author; I understood it once but forget. I do recall it is shorthand for the process of inserting a list of None for a missing key (yes, the None does need to be there).